Content Node Chain of Trust [Phase 2] - automatic node self-registration on URSM (And new Content Node Init pattern)

SidSethi opened the pull request:
### Description
_What is the purpose of this PR? What is the current behavior? New behavior? Relevant links (e.g. Trello) and/or information pertaining to PR?_

[Linear card](https://linear.app/audius/issue/AUD-189/content-node-interaction-with-userreplicasetmanager-contract-chain-of)
[Notion doc](https://www.notion.so/audiusproject/IN-PROGRESS-Chain-of-Trust-Content-Node-5c4c8f028feb40d6a893fb5c7213928d)

The PR for Phase 1 of this task is https://github.com/AudiusProject/audius-protocol/pull/1260

This is Phase 2 of chain of trust authentication flow for content nodes to register with new UserReplicaSetManager contract (URSM) in order to be eligible to be added to user's replica sets.
This PR "activates" the code from Phase 1, by adding a CN service that automatically registers the node against URSM on startup, including requesting signatures from registered nodes using the code from Phase 1 above.

Included:
- new ServiceRegistry init flow, separating L1 identity recovery, SnapbackSM init, and URSMRegistration into discrete and independent steps
- new CN Libs instance config - now has internalWeb3 logged in with DelegatePrivateKey to support chain writes (required for URSMRegistration)
- new `URSMRegistrationManager` service, responsible for registering the node. This service is entirely backwards compatible, working before and after contract is deployed
- Moves L1 identity recovery (`spID`) out of `SnapbackSM` service for better logical separation of concerns. Now handled by `ServiceRegistry`
- adding rate limit to `ursm_request_for_signature` route
- `SnapbackSM` integration test fix
- update maddog integration to test to wait for all CN to register on URSM before testing - fixes race condition in maddog CI
- new "devMode" for CN, currentlyused only for URSMRegistrationRetryTimeout, but very extensible
- Bump libs version in CN to consume required changes (also pins version)
- changing local system setup to default to 4 nodes
- new local command to update SPDelegateOwnerWallet in order to test updateDelegateOwnerWallet flow to URSM
- lots of documentation

### Tests
_List the manual tests and repro instructions to verify that this PR works as anticipated. Include log analysis if possible.\
:exclamation: If this change impacts clients, make sure that you have tested the clients :exclamation:_

1. additional validation in Mad Dog URSM test to confirm nodes were added by correct proposers
2. Manual regression testing - run client without URSM deployed and confirm user signup + track upload + track stream + replica set re-election
3. Manual testing - run client with URSM deployed and confirm user signup + track upload + track stream + replica set re-election

Testing steps:
- [x] bring up 4 nodes locally (without URSM contract deployed)
- [x] confirm maddog tests pass
- [x] test on dapp - everything should work
- [x] deploy URSM contract (without URSM self-registration turned off)
- [x] confirm maddog fails bc all nodes are not registered on URSM
- [x] test on dapp
  - [x] new user - everything should work
  - [x] old user from above - should work unless CN4 is primary
- [x] enable URSM self-registration
- [x] confirm CN4 self registers successfully
- [x] confirm maddog tests pass
- [x] test on dapp
  - [x] new user - everything should work (confirmed selected CN4 to rset)
  - [x] old user from above - everything should work
- [x] test updateDelegateOwnerWallet flow to URSM
  - [x] confirmed rate limits for new route work

Logs below show CN4 repeatedly trying to register on URSM until it succeeds
```
[2021-03-10T21:33:59.673Z]  INFO: audius_creator_node/49 on 8669656ba010: Success (requestID=wL1TLTjAy, requestMethod=GET, requestHostname=cn2_creator-node_1, requestUrl=/ursm_request_for_signature, statusCode=200, duration=438)
    requestQueryParams: spID=2&timestamp=2021-03-10T21:33:59.218Z&signature=0x742f6845195ec4bad673aae03d2dbaa49a5b599372e1c4a32ac06030847a338a1a637d0d6e7af407755baad6dc22f6e7e3d7d9dba503915890b063b93b09de5f1b
[2021-03-10T21:33:59.681Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn2_creator-node_1:4001
Executing (default): SELECT 1
[2021-03-10T21:33:59.713Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn1_creator-node_1:4000
[2021-03-10T21:33:59.741Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn4_creator-node_1:4003
Server returned error: [500] Something caused the transaction to fail for payload 6b562ad03e844d67a609adca87f9f6919320aef387e2ae123865880e746c6f99
Retry error : Error: Request failed with status code 500
Server returned error: [500] Something caused the transaction to fail for payload 6b562ad03e844d67a609adca87f9f6919320aef387e2ae123865880e746c6f99
Retry error : Error: Request failed with status code 500
Server returned error: [500] Something caused the transaction to fail for payload 6b562ad03e844d67a609adca87f9f6919320aef387e2ae123865880e746c6f99
Retry error : Error: Request failed with status code 500
[2021-03-10T21:34:15.152Z] ERROR: audius_creator_node/49 on 8669656ba010: ServiceRegistry ERROR || RegisterNodeOnL2URSM Error
[2021-03-10T21:34:25.153Z]  INFO: audius_creator_node/49 on 8669656ba010: ServiceRegistry || Attempting to register node on L2 URSM on 10000ms interval || attempt #4 ...
[2021-03-10T21:34:25.154Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Beginning URSM registration process
Executing (default): SELECT 1
Executing (default): SELECT 1
Executing (default): SELECT 1
[2021-03-10T21:34:25.506Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn1_creator-node_1:4000
[2021-03-10T21:34:25.509Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn3_creator-node_1:4002
[2021-03-10T21:34:25.539Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully received signature from http://cn4_creator-node_1:4003
[2021-03-10T21:34:25.753Z]  INFO: audius_creator_node/49 on 8669656ba010: URSMRegistrationManager || Successfully registered self on URSM
[2021-03-10T21:34:25.755Z]  INFO: audius_creator_node/49 on 8669656ba010: ServiceRegistry || URSM Registration completed
[2021-03-10T21:34:25.755Z]  INFO: audius_creator_node/49 on 8669656ba010: ServiceRegistry || All services that require server successfully initialized!
```

**:exclamation: Reminder :bulb::exclamation::**\
If this PR touches a critical flow (such as Indexing, Uploads, Gateway or the Filesystem), make sure to add the `requires-special-attention` label. **Add relevant labels as necessary.**

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,197 @@
+    const alreadyRegisteredOnURSM = !libs.Utils.isZeroAddress(
+      await libs.contracts.UserReplicaSetManagerClient.getContentNodeWallet(spID)
+    )
+    if (alreadyRegisteredOnURSM) {
```
this isn't actually true - you can overwrite the wallet parameters on chain, imagine the case where someone updates their delegatWallet on eth contracts. chain of trust should be able to rebroadcast and update with other known parties

SidSethi replied:
great point, fixed

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,197 @@
+      libs.creatorNode.serviceTypeName,
+      spID
+    )
+    if (!nodeSpInfoFromSPFactory) {
```
this will never actually return null, it will be all zero'd out - probably best to check more thoroughly here

SidSethi replied:
good call, fixed

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,197 @@
+   *    a. Confirm health check response was signed by delegate owner wallet registered on L1 for spID
+   *  5. Generate & return proposal signature artifacts
+   */
+  app.get('/ursm_request_for_proposal', handleResponse(async (req, res, next) => {
```
if this is the right url format we should standardize the other one for submit_rfp

SidSethi replied:
per other comment, that route will go away

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,197 @@
+    })
+  }))
+
+  app.get('/submitRFP', handleResponse(async (req, res, next) => {
```
why should this be a route? imagine a node coming up, requesting 3 proposer signatures - when does it have to request the node to addOrUpdate

SidSethi replied:
yeah this def won't be a route - thats why i mentioned in my msg to only look at the one `ursm_request_for_proposal` route since everything else is just testing code

hareeshnagaraj commented on creator-node/compose/env/commonEnv.sh:
```diff
@@ -1 +1,5 @@
 export CREATOR_NODE_IS_USER_METADATA=false
+export delegateOwnerWallet=0xbdd48b07f95e5925d9fbcea6b943375b8ed00fcc
```
rm these

hareeshnagaraj commented on creator-node/src/URSMService.js:
```diff
@@ -0,0 +1,234 @@
+      || LibsUtils.isZeroAddress(delegateOwnerWalletFromSPFactory)
+      || !endpointFromSPFactory
+    ) {
+      throw new Error('L1recordbad')
```
probably want to validate that the endpointFromSPFactory is actually a match to the node's configured endpoint too - remember that the spID config is always recalculated from the endpoint and is not provided by the user

hareeshnagaraj commented on contracts/contract-config.js:
```diff
@@ -4,9 +4,9 @@ module.exports = {
-    bootstrapSPIds: [],
-    bootstrapSPDelegateWallets: [],
-    bootstrapSPOwnerWallets: [],
+    bootstrapSPIds: [1,2,3],
```
remove this

hareeshnagaraj commented on creator-node/src/URSMService.js:
```diff
@@ -0,0 +1,234 @@
+    let URSMContentNodes = await this.audiusLibs.discoveryProvider.getURSMContentNodes()
+    
+    // TODO - setting list to single value for testing purposes
+    const tmp = URSMContentNodes[0]
```
remove this

hareeshnagaraj commented on creator-node/src/URSMService.js:
```diff
@@ -0,0 +1,234 @@
+      }
+
+      let nodesToAttempt = availableNodes.slice(0, numRemainingSignatures)
+      availableNodes = availableNodes.slice(numRemainingSignatures)
```
if 1/3 requests in the first iteration fails, do we move to indexes 4-6 or exclude the failed node only?

hareeshnagaraj replied:
seems like we'd want to exclude only the failures instead of requesting another 3

SidSethi replied:
discussed offline, correct as is

hareeshnagaraj commented on creator-node/src/URSMService.js:
```diff
@@ -0,0 +1,234 @@
+     * 2-a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
+     */
+    if (delegateOwnerWalletFromSPFactory === delegateOwnerWalletFromURSM) {
+      throw new Error('L2recordmatch')
```
verify ownerWallet on chain for this endpoint,delegateOwnerWallet matches the configured ownerWallet parameter in this content node (spOwnerWallet i believe)

hareeshnagaraj commented on creator-node/src/components/healthCheck/healthCheckComponentService.js:
```diff
@@ -21,10 +21,15 @@ const healthCheck = async ({ libs } = {}, logger, sequelize) => {
-  if (libs) {
-    response.selectedDiscoveryProvider = libs.discoveryProvider.discoveryProviderEndpoint
+  // If optional `randomBytesToSign` query param provided, node will include string in signed object
+  if (randomBytesToSign) {
```
why are we adding this to health_check instead of a dedicated endpoint?

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,203 @@
+   * TODO
+   * - rate limit
+   * - consider also validating L1 healthy state eg valid bounds
+   * - move to component pattern
```
component pattern should get done before we merge this i think - let's not introduce tech debt intentionally if avoidable

SidSethi replied:
this is done, in new pr

hareeshnagaraj commented on creator-node/src/routes/userReplicaSet.js:
```diff
@@ -0,0 +1,203 @@
+    /**
+     * Fetch all nodes from SPF
+     */
+    for await (const spID of [1,2,3,4]) {
```
seems like this is wrong

hareeshnagaraj replied:
whats the point of this loop?

SidSethi replied:
removed, was for dev

hareeshnagaraj commented on creator-node/src/URSMService.js:
```diff
@@ -0,0 +1,234 @@
+   *  1. Fetch node record from L1 ServiceProviderFactory for spID
+   *    a. Short-circuit if no L1 record found
+   *  2. Fetch node record from L2 UserReplicaSetManager for spID
+   *    a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
```
update description with ownerWallet matching configured spOwnerWallet

dmanjunath reviewed (commented):
The pieces all look fine. The auth scheme looks solid and the validations and edge cases seem fine. I just think this needs some polish and comments

SidSethi commented on libs/src/services/creatorNode/index.js:
```diff
@@ -2,6 +2,7 @@ const axios = require('axios')
 const FormData = require('form-data')
 
 const SchemaValidator = require('../schemaValidator')
+const { CREATOR_NODE_SERVICE_NAME } = require('./constants')
```
can remove

SidSethi commented on creator-node/src/components/replicaSet/URSMRegistrationComponentService.js:
```diff
@@ -0,0 +1,164 @@
+ *    a. Confirm health check response was signed by delegate owner wallet registered on L1 for spID
+ *  4. Generate & return proposal signature artifacts
+ */
+const respondToURSMRequestForProposal = async ({ audiusLibs, nodeConfig }, logger, spID, reqTimestamp, reqSignature) => {
```
agreed, done

SidSethi commented on creator-node/package.json:
```diff
@@ -18,7 +18,7 @@
   "license": "Apache-2.0",
   "dependencies": {
-    "@audius/libs": "^1.1.9",
+    "@audius/libs": "1.1.14",
```
need to bump again to latest before merging

SidSethi replied:
done

SidSethi commented on .circleci/config.yml:
```diff
@@ -174,7 +174,7 @@ jobs:
             sudo "$(which node)" hosts.js add
             sudo curl -L "https://github.com/docker/compose/releases/download/1.24.1/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
-            "$(which node)" setup.js up -nc 3
+            "$(which node)" setup.js up -nc 4
```
i think 4 should be our base case

hareeshnagaraj replied:
i kind of agree?

SidSethi commented on creator-node/src/snapbackSM.js:
```diff
@@ -49,21 +48,27 @@ const SyncType = Object.freeze({
 */
 class SnapbackSM {
-  constructor (audiusLibs) {
+  constructor (nodeConfig, audiusLibs) {
```
modified constructor to accept `nodeConfig` object to make this more easily testable. see code change in `test/snapbackSM.test.js`

SidSethi commented on libs/src/services/contracts/ContractClient.js:
```diff
@@ -36,7 +36,7 @@ class ContractClient {
 
   /** Inits the contract if necessary */
-  async init () {
+  async init (selectNewEndpointOnRetry = true) {
```
this is part 1 of bugfix - if we always select a new web3provider endpoint on retry, the code fails since CN is initted with only a single web3provider. also logically, this just makes more sense to not always force a provider re-selection on a contract client init op

will move to separate PR

SidSethi replied:
diff gone, on master now

SidSethi commented on libs/src/services/dataContracts/index.js:
```diff
@@ -137,7 +137,7 @@ class AudiusContracts {
     this.contracts = this.contracts || { [this.registryAddress]: 'registry' }
     this.contractAddresses = this.contractAddresses || { 'registry': this.registryAddress }
-    if (!this.contractAddresses[contractName]) {
+    if (!this.contractAddresses[contractName] || Utils.isZeroAddress(this.contractAddresses[contractName])) {
```
this is the bug that took me several hours lol

also will be moved to separate PR libs only

hareeshnagaraj replied:
fwiw i repro'd this on client by making a user prior to contract deploy

hareeshnagaraj commented on creator-node/src/services/URSMRegistrationManager.js:
```diff
@@ -0,0 +1,215 @@
+     *
+     * TODO - currently this will retry until contract is deployed, definitely not ideal
+     */
+    if (!this.audiusLibs.contracts.UserReplicaSetManagerClient) {
```
i feel like the `init` could be divorced from the `run` aspect of this flow, since we later will have to rely on a similar condition for node processing but not a blocker imo

hareeshnagaraj commented on creator-node/src/services/URSMRegistrationManager.js:
```diff
@@ -0,0 +1,215 @@
+   *    a. Error if all available nodes contacted without 3 successful signatures
+   *  5. Submit registration transaction to URSM with signatures
+   */
+  async run () {
```
i think this is a bit of a misnomer - instead of `run` this is really `init` right? since no ursm operations can happen before this completes

hareeshnagaraj replied:
oh nvm resolving this comment, for the `URSMRegistrationManager`, `run` makes total sense

hareeshnagaraj reviewed (commented):
Still reviewing rn, but do we have validation for the 'updateDelegateWallet' case?

hareeshnagaraj commented on creator-node/src/serviceRegistry.js:
```diff
@@ -15,26 +18,31 @@ const config = require('./config')
  */
 class ServiceRegistry {
   constructor () {
+    this.nodeConfig = config
```
what is the benefit of assigning to `this.nodeConfig` instead of referencing the convict object directly? if its just a reference its fine but i want to ensure we don't make a copy and end up with updates not propagating (might be overthinking this)

dmanjunath replied:
It should be a reference to the config module so updates shouldn't be a problem. But I kind of agree with @hareeshnagaraj here that we probably don't need this? I can see an argument for both adding here or not

SidSethi replied:
ya it is just a reference, same as requiring individually. am thinking of this as a configService so passing around similarly. also per my earlier comment, it makes it much much easier to unit test code

hareeshnagaraj commented on creator-node/src/components/replicaSet/replicaSetController.js:
```diff
@@ -31,6 +31,9 @@ const respondToURSMRequestForProposalController = async (req) => {
 // Routes
 
+/**
+ * TODO rate limit route (?)
```
this is a good TODO we should discuss offline

SidSethi replied:
resolved

hareeshnagaraj commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,174 @@ class ServiceRegistry {
 
-      this.snapbackSM = new SnapbackSM(this.libs)
-      await this.snapbackSM.init()
+      this._recoverNodeL1Identity()
```
instead of having each one of these spin until the previous is finished, why not have a single task that calls them in a row and just fire that off?

hareeshnagaraj replied:
not saying this is better/worse, just noticing that each of them has its own `while` clause that spins until the prior condition is completed

SidSethi replied:
way cleaner, fixed

hareeshnagaraj commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,174 @@ class ServiceRegistry {
+    while (complete === false) {
+      try {
+        this.snapbackSM = new SnapbackSM(this.nodeConfig, this.libs)
+        await this.snapbackSM.init()
```
do these two things actually relate to each other? also can't seem to find the snapbackSM.init() function - where is that found?

SidSethi replied:
good catch, the constructor should have been called outside loop, fixed

hareeshnagaraj reviewed (changes requested):
Some points of clarification

SidSethi commented on creator-node/package.json:
```diff
@@ -18,7 +18,7 @@
   "license": "Apache-2.0",
   "dependencies": {
-    "@audius/libs": "^1.1.9",
+    "@audius/libs": "1.1.16",
```
note this also pins version

SidSethi commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,198 @@ class ServiceRegistry {
+      web3Config: AudiusLibs.configInternalWeb3(
+        config.get('dataRegistryAddress'),
+        [config.get('dataProviderUrl')],
+        // TODO - formatting this private key here is not ideal
```
cc @hareeshnagaraj

hareeshnagaraj replied:
unf idt we have an option here :/

SidSethi replied:
eh, i think there's a cleaner way to do this but will leave for now

hareeshnagaraj commented on creator-node/src/config.js:
```diff
@@ -183,6 +183,12 @@ const config = convict({
+    doc: 'Total requests per hour rate limit for /ursm_request_for_signature route',
+    format: 'nat',
+    env: 'URSMRequestForSignatureReqLimit',
+    default: null
```
should this be defaulted to null?

SidSethi replied:
just went with the established pattern for the rate limit configs, which are all defaulted to null

SidSethi replied:
yeh could have done it that way, would have maybe been better, just stuck with current pattern without too much thought tbh

hareeshnagaraj commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,197 @@ class ServiceRegistry {
+    }
+
+    // Cannot progress without recovering spID from node's record on L1 ServiceProviderFactory contract
+    await this._recoverNodeL1Identity()
```
what happens if one of these fails?

hareeshnagaraj replied:
i see, each is internally retried indefinitely

SidSethi replied:
yee

SidSethi replied:
added a comment tho

dmanjunath commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,197 @@ class ServiceRegistry {
+    while (true) {
+      this.logInfo(`Attempting to init UserReplicaSetManagerClient on ${retryTimeoutMs}ms interval...`)
+      try {
+        await this.libs.contracts.initUserReplicaSetManagerClient(false)
```
if i understand correctly, this line tries to init the client, the if condition checks after the init if the client exists and the break is just to exit this while loop if it doesn't exist yet

SidSethi replied:
the purpose of putting this here is so we don't do an unnecessary timeout and wait for the next loop iteration before exiting

draj you're correct except that the exit condition is not "if it doesn't existyet" but rather if it **does** exist

hareeshnagaraj replied:
yeah this is actually a no-op if the client has been registered successfully so i think we have no isse here

dmanjunath commented on creator-node/src/serviceRegistry.js:
```diff
@@ -44,64 +52,197 @@ class ServiceRegistry {
+  async _registerNodeOnL2URSM () {
+    // Wait until URSM contract has been deployed (for backwards-compatibility)
+    let retryTimeoutMs = (this.nodeConfig.get('devMode')) ? 10000 /** 10sec */ : 600000 /* 10min */
+    while (true) {
```
can we change these while(true) to like while(!initialized) and set initialized whenever this flow is done? while true just makes me nervous

SidSethi replied:
hahaha i had a feeling you would say this - ya sure, changing

dmanjunath reviewed (commented):
this PR looks great. had one comment about while true. i'm basically ready to approve if you change that or feel we shouldn't change that

SidSethi commented:
@dmanjunath @hareeshnagaraj addressed all comments pls re-review
also CI will bring up all these and register everything so am trusting that to test my last code change instead of running whole thing locally, lmk if you want me to test locally

hareeshnagaraj commented on mad-dog/src/tests/test_userReplicaSetManager.js:
```diff
@@ -230,8 +255,29 @@ const verifyUrsmContentNodes = async (executeOne) => {
       logger.info(`Found UserReplicaSetManager and ServiceProviderFactory match for spID=${spID}, delegateWallet=${delegateWalletFromChain}, ownerWallet=${ownerWalletFromChain}`)
+
+      /**
+       * Confirm nodes with spIDs 1-3 are correctly configured as bootstrappers and any additional nodes were registered with proposals from registered nodes
```
just to clarify, we know the order here because ursm content nodes are ordered by ID in the route from disc prov and the requesting node aggregates the proposer responses in said order?

SidSethi replied:
nah we don't know the order in which the requesting node aggregates proposer responses. this code works bc it makes no assumptions on the order of those responses

hareeshnagaraj reviewed (commented):
Changes look great, love the code quality and thoroughness here - comments, testing, PR description 💯

hareeshnagaraj reviewed (approved):
LGTM!

dmanjunath reviewed (approved):
YESSSSSSS
